packages/apollo-server-core/src/__tests__/dataSources.test.ts

```diff
@@ -0,0 +1,96 @@
+import { ApolloServerBase } from '../ApolloServer';
+import gql from 'graphql-tag';
+
+const typeDefs = gql`
+  type Query {
+    hello: String
+  }
+`;
+
+const resolvers = {
+  Query: {
+    hello() {
+      return 'world';
+    }
+  },
+};
+
+describe('ApolloServerBase dataSources', () => {
+  it('initializes synchronous datasources from a datasource creator function', async () => {
+    const initialize = jest.fn();
+
+    const server = new ApolloServerBase({
+      typeDefs,
+      resolvers,
+      dataSources: () => ({ x: { initialize }, y: { initialize } })
+    });
+
+    await server.executeOperation({ query: "query { hello }"});
+
+    expect(initialize).toHaveBeenCalledTimes(2);
+  });
+
+  it('initializes asynchronous datasources before calling resolvers', async () => {
+    const expectedMessage = 'success';
+    let maybeInitialized = 'failure';
+
+    const additionalInitializer = jest.fn();
+
+    const server = new ApolloServerBase({
+      typeDefs,
+      resolvers: {
+        Query: {
+          hello(_, __, context) {
+            return context.dataSources.x.getData();
+          }
+        },
+      },
+      dataSources: () => ({
+        x: {
+          initialize() {
+            return new Promise(res => { setTimeout(() => {
+              maybeInitialized = expectedMessage;
+              res();
+            }, 200) })
+          },
+          getData() { return maybeInitialized; }
+        },
+        y: {
+          initialize() {
+            return new Promise(res => { setTimeout(() => {
+              additionalInitializer();
+              res();
+            }, 400) })
+          }
+        }
+      })
+    });
+
```

Comment:
To "sleep" quickly during tests, Jest provides a few timer mocking functions so as not to introduce long delays due to timers. Would you mind using these for this test?

`jest.useFakeTimers()` - mocks timers
`jest.advanceTimersByTime(400)` - fast-forwards timers by a specified ms
`jest.useRealTimers()` - unmocks timers

[Complete Timer Mocks Reference](https://jestjs.io/docs/en/jest-object#mock-timers)